File: crates/tauri-plugin-gns/guest-js/src/trust.ts
/**
 * @file GNS Trust System
 * @description Trust score calculation and identity verification via Proof-of-Trajectory
 * @module @anthropic/tauri-plugin-gns-api/trust
 */

import { invoke } from '@tauri-apps/api/core';
import type {
  TrustScore,
  TrustTier,
  TrustVerification,
  TrustRequirements,
  TrustComponents,
} from './types';

/**
 * Get the trust score for the active identity.
 * 
 * Trust score is calculated from:
 * - Trajectory quality (pattern consistency)
 * - Temporal consistency (regular collection)
 * - Chain integrity (valid hash chains)
 * - Epoch reliability (published epochs)
 * - Geographic diversity (unique locations)
 * 
 * @example
 * ```typescript
 * const score = await getTrustScore();
 * console.log(`Trust: ${score.score}% (${score.tier})`);
 * ```
 * 
 * @returns Trust score with breakdown
 */
export async function getTrustScore(): Promise<TrustScore> {
  return invoke<TrustScore>('plugin:gns|get_trust_score');
}

/**
 * Get detailed trust score with full component breakdown.
 * 
 * @example
 * ```typescript
 * const details = await getTrustDetails();
 * console.log(`Trajectory Quality: ${details.components.trajectoryQuality}%`);
 * console.log(`Geographic Diversity: ${details.components.geographicDiversity}%`);
 * ```
 * 
 * @returns Complete trust score analysis
 */
export async function getTrustDetails(): Promise<TrustScore> {
  return invoke<TrustScore>('plugin:gns|get_trust_details');
}

/**
 * Verify an identity meets specified trust requirements.
 * 
 * @example
 * ```typescript
 * const verification = await verifyIdentity('abc123...', {
 *   minTrustScore: 40,
 *   minBreadcrumbs: 200,
 * });
 * 
 * if (verification.isVerified) {
 *   console.log('Identity meets requirements');
 * } else {
 *   verification.checks.filter(c => !c.passed).forEach(c => {
 *     console.log(`Failed: ${c.name} - ${c.details}`);
 *   });
 * }
 * ```
 * 
 * @param publicKey - Identity to verify
 * @param requirements - Trust requirements to check
 * @returns Verification result with individual check results
 */
export async function verifyIdentity(
  publicKey?: string,
  requirements?: TrustRequirements
): Promise<TrustVerification> {
  return invoke<TrustVerification>('plugin:gns|verify_identity', {
    publicKey: publicKey ?? null,
    requirements: requirements ?? null,
  });
}

/**
 * Get the display name for a trust tier.
 * 
 * @example
 * ```typescript
 * getTierDisplayName('Rooted'); // 'Rooted'
 * ```
 */
export function getTierDisplayName(tier: TrustTier): string {
  return tier;
}

/**
 * Get the emoji for a trust tier.
 * 
 * @example
 * ```typescript
 * getTierEmoji('Verified'); // '⭐'
 * ```
 */
export function getTierEmoji(tier: TrustTier): string {
  const emojis: Record<TrustTier, string> = {
    Seedling: '🌱',
    Rooted: '🌿',
    Established: '🌲',
    Trusted: '🏔️',
    Verified: '⭐',
  };
  return emojis[tier];
}

/**
 * Get the color for a trust tier (hex).
 * 
 * @example
 * ```typescript
 * getTierColor('Trusted'); // '#3b82f6'
 * ```
 */
export function getTierColor(tier: TrustTier): string {
  const colors: Record<TrustTier, string> = {
    Seedling: '#9ca3af',   // Gray
    Rooted: '#22c55e',     // Green
    Established: '#3b82f6', // Blue
    Trusted: '#a855f7',    // Purple
    Verified: '#f59e0b',   // Gold
  };
  return colors[tier];
}

/**
 * Get the minimum score for a trust tier.
 * 
 * @example
 * ```typescript
 * getTierMinScore('Rooted'); // 20
 * ```
 */
export function getTierMinScore(tier: TrustTier): number {
  const minScores: Record<TrustTier, number> = {
    Seedling: 0,
    Rooted: 20,
    Established: 40,
    Trusted: 60,
    Verified: 80,
  };
  return minScores[tier];
}

/**
 * Determine trust tier from a score.
 * 
 * @example
 * ```typescript
 * getTierFromScore(45); // 'Established'
 * ```
 */
export function getTierFromScore(score: number): TrustTier {
  if (score >= 80) return 'Verified';
  if (score >= 60) return 'Trusted';
  if (score >= 40) return 'Established';
  if (score >= 20) return 'Rooted';
  return 'Seedling';
}

/**
 * Get preset trust requirements for handle claiming.
 * 
 * @returns Requirements for claiming a @handle
 */
export function getHandleClaimRequirements(): TrustRequirements {
  return {
    minTrustScore: 20,
    minBreadcrumbs: 100,
    minAccountAgeDays: 7,
    minUniqueLocations: 10,
    requiredTier: 'Rooted',
  };
}

/**
 * Get preset trust requirements for payments.
 * 
 * @returns Requirements for sending/receiving payments
 */
export function getPaymentRequirements(): TrustRequirements {
  return {
    minTrustScore: 40,
    minBreadcrumbs: 200,
    minAccountAgeDays: 14,
    minUniqueLocations: 20,
    requiredTier: 'Established',
  };
}

/**
 * Get empty trust requirements (no verification needed).
 * 
 * @returns Empty requirements object
 */
export function getNoRequirements(): TrustRequirements {
  return {};
}

/**
 * Check if an identity meets handle claim requirements.
 * 
 * @example
 * ```typescript
 * if (await canClaimHandle()) {
 *   await claimHandle('myhandle');
 * }
 * ```
 */
export async function canClaimHandle(publicKey?: string): Promise<boolean> {
  const verification = await verifyIdentity(publicKey, getHandleClaimRequirements());
  return verification.isVerified;
}

/**
 * Check if an identity meets payment requirements.
 * 
 * @example
 * ```typescript
 * if (await canSendPayments()) {
 *   await sendPayment('@alice', 10);
 * }
 * ```
 */
export async function canSendPayments(publicKey?: string): Promise<boolean> {
  const verification = await verifyIdentity(publicKey, getPaymentRequirements());
  return verification.isVerified;
}

/**
 * Format a trust score for display.
 * 
 * @example
 * ```typescript
 * formatTrustScore(score); // '45% 🌲 Established'
 * ```
 */
export function formatTrustScore(score: TrustScore): string {
  return `${score.score}% ${getTierEmoji(score.tier)} ${score.tier}`;
}

/**
 * Get a human-readable description of trust components.
 * 
 * @example
 * ```typescript
 * const desc = describeTrustComponents(score.components);
 * console.log(desc.trajectoryQuality); // 'Strong trajectory patterns detected'
 * ```
 */
export function describeTrustComponents(components: TrustComponents): Record<keyof TrustComponents, string> {
  const describe = (value: number, lowDesc: string, midDesc: string, highDesc: string): string => {
    if (value < 30) return lowDesc;
    if (value < 70) return midDesc;
    return highDesc;
  };

  return {
    trajectoryQuality: describe(
      components.trajectoryQuality,
      'Trajectory patterns are still developing',
      'Good trajectory patterns detected',
      'Strong trajectory patterns detected'
    ),
    temporalConsistency: describe(
      components.temporalConsistency,
      'Breadcrumb collection is irregular',
      'Regular breadcrumb collection',
      'Excellent collection consistency'
    ),
    chainIntegrity: describe(
      components.chainIntegrity,
      'Hash chain has some gaps',
      'Hash chain is mostly complete',
      'Perfect hash chain integrity'
    ),
    epochReliability: describe(
      components.epochReliability,
      'Few epochs published',
      'Good epoch publishing rate',
      'Excellent epoch publishing'
    ),
    geographicDiversity: describe(
      components.geographicDiversity,
      'Limited geographic coverage',
      'Good geographic diversity',
      'Excellent geographic diversity'
    ),
  };
}

/**
 * Calculate days until reaching a target trust tier.
 * 
 * This is an estimate based on current progress rate.
 * 
 * @example
 * ```typescript
 * const days = estimateDaysToTier(score, 'Trusted');
 * console.log(`About ${days} days to reach Trusted tier`);
 * ```
 */
export function estimateDaysToTier(
  currentScore: TrustScore,
  targetTier: TrustTier
): number | null {
  const targetMin = getTierMinScore(targetTier);
  
  if (currentScore.score >= targetMin) {
    return 0; // Already at or above target
  }
  
  if (currentScore.accountAgeDays === 0) {
    return null; // Cannot estimate without history
  }
  
  // Estimate rate of score increase per day
  const dailyRate = currentScore.score / currentScore.accountAgeDays;
  
  if (dailyRate <= 0) {
    return null; // No progress
  }
  
  const scoreDiff = targetMin - currentScore.score;
  return Math.ceil(scoreDiff / dailyRate);
}

File: crates/tauri-plugin-gns/src/models/trust.rs
//! Trust Models
//!
//! Trust scoring based on Proof-of-Trajectory.
//! Trust is earned through physical presence, not purchased.

use serde::{Deserialize, Serialize};

/// Trust score details
///
/// Trust is calculated from breadcrumb patterns and cannot be faked
/// without physical presence across diverse locations over time.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrustScore {
    /// Overall trust score (0-100)
    pub score: f64,

    /// Trust tier (Seedling → Rooted → Established → Trusted → Verified)
    pub tier: TrustTier,

    /// Component scores
    pub components: TrustComponents,

    /// When the score was calculated
    pub calculated_at: String,

    /// Total breadcrumb count
    pub breadcrumb_count: u32,

    /// Account age in days
    pub account_age_days: u32,

    /// Number of unique H3 cells visited
    pub unique_locations: u32,

    /// Number of published epochs
    pub epoch_count: u32,
}

/// Trust score component breakdown
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrustComponents {
    /// Trajectory quality (spatial diversity)
    pub trajectory_quality: f64,

    /// Temporal consistency (regular activity)
    pub temporal_consistency: f64,

    /// Chain integrity (unbroken breadcrumb chain)
    pub chain_integrity: f64,

    /// Epoch reliability (published proofs)
    pub epoch_reliability: f64,

    /// Geographic diversity
    pub geographic_diversity: f64,
}

/// Trust tiers based on score ranges
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum TrustTier {
    /// 0-19: New identity, minimal trajectory
    Seedling,
    /// 20-39: Building trajectory
    Rooted,
    /// 40-59: Established presence
    Established,
    /// 60-79: Trusted identity
    Trusted,
    /// 80-100: Fully verified through extensive trajectory
    Verified,
}

impl TrustTier {
    /// Get the tier from a score
    pub fn from_score(score: f64) -> Self {
        match score as u32 {
            0..=19 => TrustTier::Seedling,
            20..=39 => TrustTier::Rooted,
            40..=59 => TrustTier::Established,
            60..=79 => TrustTier::Trusted,
            _ => TrustTier::Verified,
        }
    }

    /// Get the display name
    pub fn display_name(&self) -> &'static str {
        match self {
            TrustTier::Seedling => "Seedling",
            TrustTier::Rooted => "Rooted",
            TrustTier::Established => "Established",
            TrustTier::Trusted => "Trusted",
            TrustTier::Verified => "Verified",
        }
    }

    /// Get the tier color (for UI)
    pub fn color(&self) -> &'static str {
        match self {
            TrustTier::Seedling => "#94a3b8",    // gray
            TrustTier::Rooted => "#84cc16",      // lime
            TrustTier::Established => "#22c55e", // green
            TrustTier::Trusted => "#0ea5e9",     // sky blue
            TrustTier::Verified => "#8b5cf6",    // purple
        }
    }

    /// Get the emoji for the tier
    pub fn emoji(&self) -> &'static str {
        match self {
            TrustTier::Seedling => "🌱",
            TrustTier::Rooted => "🌿",
            TrustTier::Established => "🌲",
            TrustTier::Trusted => "🏔️",
            TrustTier::Verified => "⭐",
        }
    }

    /// Minimum score for this tier
    pub fn min_score(&self) -> f64 {
        match self {
            TrustTier::Seedling => 0.0,
            TrustTier::Rooted => 20.0,
            TrustTier::Established => 40.0,
            TrustTier::Trusted => 60.0,
            TrustTier::Verified => 80.0,
        }
    }
}

/// Trust verification result
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrustVerification {
    /// The identity that was verified
    pub identity: String,

    /// Whether the identity meets minimum requirements
    pub is_verified: bool,

    /// Current trust score
    pub trust_score: TrustScore,

    /// Verification checks performed
    pub checks: Vec<TrustCheck>,

    /// When verification was performed
    pub verified_at: String,
}

/// Individual trust verification check
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrustCheck {
    /// Check name
    pub name: String,

    /// Whether check passed
    pub passed: bool,

    /// Check details
    pub details: String,

    /// Required value (if applicable)
    pub required: Option<String>,

    /// Actual value (if applicable)
    pub actual: Option<String>,
}

/// Requirements for different operations
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrustRequirements {
    /// Minimum trust score
    pub min_trust_score: f64,

    /// Minimum breadcrumb count
    pub min_breadcrumbs: u32,

    /// Minimum account age in days
    pub min_account_age_days: u32,

    /// Minimum unique locations
    pub min_unique_locations: u32,

    /// Required tier
    pub required_tier: Option<TrustTier>,
}

impl Default for TrustRequirements {
    fn default() -> Self {
        Self {
            min_trust_score: 20.0,
            min_breadcrumbs: 100,
            min_account_age_days: 7,
            min_unique_locations: 10,
            required_tier: Some(TrustTier::Rooted),
        }
    }
}

impl TrustRequirements {
    /// Requirements for claiming a handle
    pub fn for_handle_claim() -> Self {
        Self {
            min_trust_score: 20.0,
            min_breadcrumbs: 100,
            min_account_age_days: 7,
            min_unique_locations: 10,
            required_tier: Some(TrustTier::Rooted),
        }
    }

    /// Requirements for sending payments
    pub fn for_payment() -> Self {
        Self {
            min_trust_score: 40.0,
            min_breadcrumbs: 200,
            min_account_age_days: 14,
            min_unique_locations: 20,
            required_tier: Some(TrustTier::Established),
        }
    }

    /// No requirements (for basic messaging)
    pub fn none() -> Self {
        Self {
            min_trust_score: 0.0,
            min_breadcrumbs: 0,
            min_account_age_days: 0,
            min_unique_locations: 0,
            required_tier: None,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_trust_tier_from_score() {
        assert_eq!(TrustTier::from_score(0.0), TrustTier::Seedling);
        assert_eq!(TrustTier::from_score(19.9), TrustTier::Seedling);
        assert_eq!(TrustTier::from_score(20.0), TrustTier::Rooted);
        assert_eq!(TrustTier::from_score(50.0), TrustTier::Established);
        assert_eq!(TrustTier::from_score(75.0), TrustTier::Trusted);
        assert_eq!(TrustTier::from_score(100.0), TrustTier::Verified);
    }

    #[test]
    fn test_tier_colors() {
        assert!(!TrustTier::Verified.color().is_empty());
        assert!(!TrustTier::Seedling.emoji().is_empty());
    }

    #[test]
    fn test_requirements() {
        let handle_req = TrustRequirements::for_handle_claim();
        assert_eq!(handle_req.min_breadcrumbs, 100);

        let payment_req = TrustRequirements::for_payment();
        assert!(payment_req.min_trust_score > handle_req.min_trust_score);
    }
}
